cli/download_source.go

```diff
@@ -356,9 +365,38 @@ func getTerraformSourceUrl(terragruntOptions *options.TerragruntOptions, terragr
 	}
 }
 
-// Download the code from the Canonical Source URL into the Download Folder using the terraform init command
-func terraformInit(terraformSource *TerraformSource, terragruntOptions *options.TerragruntOptions, terragruntConfig *config.TerragruntConfig) error {
-	terragruntOptions.Logger.Printf("Downloading Terraform configurations from %s into %s using terraform init", terraformSource.CanonicalSourceURL, terraformSource.DownloadDir)
+// Download the code from the Canonical Source URL into the Download Folder using the go-getter library
+func downloadSource(terraformSource *TerraformSource, terragruntOptions *options.TerragruntOptions, terragruntConfig *config.TerragruntConfig) error {
+	terragruntOptions.Logger.Printf("Downloading Terraform configurations from %s into %s", terraformSource.CanonicalSourceURL, terraformSource.DownloadDir)
+
+	// go-getter will not download into folders that already exist, so initially, we download into a brand new temp
+	// folder
+	if err := os.MkdirAll(terragruntOptions.DownloadDir, 0700); err != nil {
+		return errors.WithStackTrace(err)
+	}
+	tmpDownloadDir, err := ioutil.TempDir(terragruntOptions.DownloadDir, "terragrunt-download-temp")
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	if err := os.RemoveAll(tmpDownloadDir); err != nil {
+		return errors.WithStackTrace(err)
+	}
+	if err := getter.GetAny(tmpDownloadDir, terraformSource.CanonicalSourceURL.String()); err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	// Now copy all the contents of the tmp folder into the original download dir
+	if err := os.MkdirAll(terraformSource.DownloadDir, 0700); err != nil {
+		return errors.WithStackTrace(err)
+	}
+	if err := util.CopyFolderContents(tmpDownloadDir, terraformSource.DownloadDir); err != nil {
+		return err
+	}
+
+	// Clean up the tmp folder
+	if err := os.RemoveAll(tmpDownloadDir); err != nil {
+		return errors.WithStackTrace(err)
+	}
```

Comment:
Why not use `defer` here?

Reply:
Done: https://github.com/gruntwork-io/terragrunt/pull/731/commits/e2398083111a4c1c3d47dbf34ca7154f6ae6bcf3